x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SearchResponse, SearchHit, SignalHit } from '../../types';
+import { Logger } from '../../../../../../../../src/core/server';
+import { AlertServices } from '../../../../../alerting/server/types';
+
+export class ScrollAndBulkIndex {
+  service: AlertServices;
+  logger: Logger;
+  signalParams: Record<string, any>; // eslint-disable-line @typescript-eslint/no-explicit-any
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  constructor(service: AlertServices, params: Record<string, any>, logger: Logger) {
+    this.service = service;
+    this.logger = logger;
+    this.signalParams = params;
+  }
```

Comment:
This is a clean class and follows really great practices 👍which makes me sad to put a "however" below. :-(

However, I would say that there is an almost complete reduction of uses of classes across the back NodeJS side as well as the front end. Most of the reduction is because of newer frameworks such as ReactJS, Redux, ts-fp, lodash/fp, RxJS, etc... which all embrace a mostly pure function approach.

Most code is being refactored into functions which takes their dependencies as arguments and mostly just work as pure functions. 

For example, the `buildBulkBody` could just be something like:

`buildBulkBody(doc, signalParams, logger)`

where all its dependencies are just passed in. The only dependencies that don't have to be passed in would be other pure functions as they are pure. 

A positive set of natural properties you gain if you were to break out all the functions into pure functions and don't use a class structure as you grow the code base are these:

* If any particular function grows too large or if your file with all your functions grows too large in size to fit all the code on a single screen, you can re-locate the functions to new files and not have to worry about anything breaking. Where with a class, you have to break it apart into several smaller interdependent classes and then create multiple instances and wire them together in an object graph.

* When other maintainers come into your code they cannot easily create derived state and add it to the `this` context and begin having odd state issues that you might have to fix ;-) Instead, they have to modify arguments to add the new state which is a good thing. It pushes state management back up the call graph which is what we will have to maintain rather than an object graph.

* Testing is easier as you don't have to create new instances of this class and set it up. You can just test each function separate.

* As our state grows within the call stack, it is much easier to understand when a function is called where the state originates as we can look up the call stack. In a class based approach we have to look at each member function to see if something could have modified a piece of state we are dependent on which can lead to a mental combinatorial tracking effect as we grow the functions.     

* And finally, if I see an interesting function and want to use it, I don't have to go up the call-graph and try to pass down the whole object instance with sets of functions I don't want through multiple layers to call a function I might have all the dependencies for already. Likewise, calling object functions could introduce more side-effects and issues if state is managed is accidentally managed there.

If for some reason I'm missing a dependency deep in my call-graph to call the function that is actually a good problem as either I have to pass down the dependency through my functions in the call graph to make it work, or I have to think if up above I really want to call the function and create a derived piece of state to pass down.

Reply:
This is awesome feedback! Thanks I agree an fp approach is a better way to manage this stuff and will definitely allow for easier testing. I'll rework this 👍